File: fig2.py
import numpy as np
import matplotlib.pyplot as plt
from NSFEM import NSFEM
from contour import contour

ns = NSFEM('obstacle/node.txt',
           'obstacle/element.txt',
           'obstacle/stream_fix.txt',
           'obstacle/vortex_fix.txt',
           'obstacle/vortex_wall.txt')
region = np.loadtxt('obstacle/region.txt')

n = ns.run(0.2, 0.05, 0.01)
print(n)

plt.figure(figsize=(5,4.5))

x,y = ns.node.T
X,Y = region.T

plt.subplot(311)
plt.axis('equal')
plt.axis([0,20,0,5])
ns.plot_velocity()
plt.plot(X, Y, 'k')
plt.box('off')
plt.xticks(())
plt.yticks(())

plt.subplot(312)
plt.axis('equal')
plt.axis([0,20,0,5])
contour(x, y, ns.stream, 16, confine=region, cmap='jet')
plt.plot(X, Y, 'k')
plt.box('off')
plt.xticks(())
plt.yticks(())

plt.subplot(313)
plt.axis('equal')
plt.axis([0,20,0,5])
contour(x, y, ns.vortex, 32, confine=region, cmap='jet')
plt.plot(X, Y, 'k')
plt.box('off')
plt.xticks(())
plt.yticks(())

plt.tight_layout()
plt.savefig('fig2.eps')
plt.show()

File: C++/fig2.cpp
#include<iostream>
#include<fstream>
#include "NSFEM.h"
using namespace NSFEM;

main() {
    int i;
    double x,y;
    std::ofstream f;

    init("../obstacle/node.txt",
         "../obstacle/element.txt",
         "../obstacle/stream_fix.txt",
         "../obstacle/vortex_fix.txt",
         "../obstacle/vortex_wall.txt");
    i = run(0.2, 0.05, 0.01);
    std::cout << i << std::endl;

    f.open("fig2uv.txt");
    for(i=0; i<element.nrows(); i++) {
        elem_center(x, y, i);
        f << x << ' ';
        f << y << ' ';
        f << velocity[i][0] << ' ';
        f << velocity[i][1] << '\n';
    }
    f.close();

    f.open("fig2ps.txt");
    for(i=0; i<node.nrows(); i++) {
        f << node[i][0] << ' ';
        f << node[i][1] << ' ';
        f << stream[i] << '\n';
    }
    f.close();

    f.open("fig2om.txt");
    for(i=0; i<node.nrows(); i++) {
        f << node[i][0] << ' ';
        f << node[i][1] << ' ';
        f << vortex[i] << '\n';
    }
    f.close();
}
